### PR TITLE
Add Package version details and approval fields

### PR DESCRIPTION
**Pull request by praveenkuttappan**
- Added package name, package version and approval fields in review file and revision models.
- Package name and package version is set from C# language processors.

### CODE REVIEW

**Comment by praveenkuttappan** [comment]
@JonathanGiles : FYI.. This PR has changes to add language and other fields in review

**Comment by JonathanGiles** [comment]
@praveenkuttappan I note that you've updated the model version to 19, so I will do that for Java too. I'm curious if you need me to add support for approval fields in the json, and if so, what fields and what is expected?

**Comment by praveenkuttappan** [comment]
@JonathanGiles  Approval fields are not required in language specific code files. It's stored in revision metadata in cosmos DB.

**Comment by pakrym** [comment]
Hm, I'm not a fan of adding these properties without having any consumers for them. Let's instead add them as part of implementing the features they would be a part of.

**Comment by praveenkuttappan** [comment]
> Hm, I'm not a fan of adding these properties without having any consumers for them. Let's instead add them as part of implementing the features they would be a part of.

I got it. Reason for adding these properties is to start getting these properties updated by language processors. My plan was implement them in chunks rather than a a big functionality. I will check and see if I can pull in more changes that uses these properties.

**Comment by pakrym** [comment]
>Reason for adding these properties is to start getting these properties updated by language processors

I'm OK with this one if we limit it to properties actually being populated like language/version etc. But we also need to decide what some of them mean across languages.

**Comment by praveenkuttappan** [comment]
> > Reason for adding these properties is to start getting these properties updated by language processors
> 
> I'm OK with this one if we limit it to properties actually being populated like language/version etc. But we also need to decide what some of them mean across languages.
> 
> Approval related properties are not generator-provided.

I have reverted changes and also removed  packageVersion field as we discussed.

**Comment by praveenkuttappan** [comment]
@JonathanGiles : Can you please remove PackageVersion from Codefile in java code generator. Only PackageName needs to be filled by JAva processor. Language is anyway added by APIView tool when we process review file from java parser.

**Comment by praveenkuttappan** [comment]
@pakrym @JonathanGiles : Here is the first version of Language filter UI change as part of this PR. 
image

**Comment by JonathanGiles** [comment]
I'm not clear on what you want me to support. At present I have added language, package name and package version. Do you want me to just disable package version for now and add it back later, or do you not think you need it now? You still need language, correct?

**Comment by praveenkuttappan** [comment]
> I'm not clear on what you want me to support. At present I have added language, package name and package version. Do you want me to just disable package version for now and add it back later, or do you not think you need it now? You still need language, correct?

You can just remove PackageVersion field. We don’t need it at the moment.

File: src/dotnet/APIView/APIViewWeb/Repositories/CosmosReviewRepository.cs [hunk 1]
```diff
@@ -33,6 +33,10 @@ public async Task<IEnumerable<ReviewModel>> GetReviewsAsync(bool closed)
                 allReviews.AddRange(result.Resource);
             }
 
+            if(!string.IsNullOrEmpty(Language) && Language != "All")
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Repositories/CosmosReviewRepository.cs:36]: we should filter this as part of the query same as we do with IsClosed.

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs [hunk 2]
```diff
@@ -60,9 +60,9 @@ public async Task<ReviewModel> CreateReviewAsync(ClaimsPrincipal user, string or
             return review;
         }
 
-        public Task<IEnumerable<ReviewModel>> GetReviewsAsync(bool closed)
+        public Task<IEnumerable<ReviewModel>> GetReviewsAsync(bool closed, string Language)
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:63]:
```suggestion
        public Task<IEnumerable<ReviewModel>> GetReviewsAsync(bool closed, string language)
```

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs [hunk 3]
```diff
@@ -255,6 +255,7 @@ private void InitializeFromCodeFile(ReviewCodeFileModel file, CodeFile codeFile)
             file.Language = codeFile.Language;
             file.VersionString = codeFile.VersionString;
             file.Name = codeFile.Name;
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:257]: How is name different from the package name?

File: src/dotnet/APIView/APIViewWeb/Models/ReviewModel.cs [hunk 4]
```diff
@@ -58,7 +58,7 @@ public bool IsUserSubscribed(ClaimsPrincipal user)
 
         public string GetUserEmail(ClaimsPrincipal user) =>
             NotificationManager.GetUserEmail(user);
-
+
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Models/ReviewModel.cs:61]: nit: ws

File: src/dotnet/APIView/APIViewWeb/Repositories/CosmosReviewRepository.cs (hunk 1 above)
Comment (praveenkuttappan) [reply src/dotnet/APIView/APIViewWeb/Repositories/CosmosReviewRepository.cs:36]: Language is not saved in Review model currently. All of existing records has language in revisionmodel which is child of review. But you are right. We need packagename in ReviewModel itself to query and find current master version is approved or not.

Comment (pakrym) [reply src/dotnet/APIView/APIViewWeb/Repositories/CosmosReviewRepository.cs:36]: I think we can use the query for inner objects as well. Cosmos query language should allow it.

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs (hunk 3 above)
Comment (praveenkuttappan) [reply src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:257]: Currently Name is package name and version name. Package Name is to standardize this to have only package name for lookup purpose when autodetecting API difference for a package is added.

File: src/dotnet/APIView/APIViewWeb/Repositories/CosmosReviewRepository.cs (hunk 1 above)
Comment (praveenkuttappan) [reply src/dotnet/APIView/APIViewWeb/Repositories/CosmosReviewRepository.cs]: I have fixed this to refer inner variables using Exists clause

File: src/dotnet/APIView/APIViewWeb/Models/ReviewModel.cs [hunk 5]
```diff
@@ -75,6 +75,9 @@ public string DisplayName
         public DateTime LastUpdated => Revisions.LastOrDefault()?.CreationDate ?? CreationDate;
 
         [JsonIgnore]
-        public string Language => Revisions.FirstOrDefault()?.Files.FirstOrDefault()?.Language;
+        public string Language => Revisions.LastOrDefault()?.Files.LastOrDefault()?.Language;
+
+        [JsonIgnore]
+        public string PackageName => Revisions.LastOrDefault()?.Files.LastOrDefault()?.PackageName;
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Models/ReviewModel.cs:81]: Are these two used?

**Comment by pakrym** [comment]
Can you deploy to staging for a final look please?

File: src/dotnet/APIView/APIViewWeb/Models/ReviewModel.cs (hunk 5 above)
Comment (praveenkuttappan) [reply src/dotnet/APIView/APIViewWeb/Models/ReviewModel.cs]: Reverted this change. Not required anymore in current PR.

**Comment by praveenkuttappan** [comment]
This change is deployed on staging instance now. Please note that there are few json reviews incorrectly tagged as C# which should get fixed when they upload new version of wheel going forward.

**Review by pakrym — APPROVED** [review]
Great job!

**Pull request merged**